internal/copy/copy_dir.go

```diff
@@ -92,34 +92,48 @@ func CopyDir(dst, src string) error {
 		if info.Mode()&os.ModeSymlink == os.ModeSymlink {
 			target, err := os.Readlink(path)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to read symlink %q: %w", path, err)
 			}
 
-			return os.Symlink(target, dstPath)
+			if err := os.Symlink(target, dstPath); err != nil {
+				return fmt.Errorf("failed to create symlink %q: %w", dstPath, err)
+			}
+			return nil
 		}
 
-		// If we have a file, copy the contents.
-		srcF, err := os.Open(path)
-		if err != nil {
+		if err := copyFile(dstPath, path, info.Mode()); err != nil {
 			return err
 		}
-		defer srcF.Close()
 
-		dstF, err := os.Create(dstPath)
-		if err != nil {
-			return err
-		}
-		defer dstF.Close()
+		return nil
+	}
 
-		if _, err := io.Copy(dstF, srcF); err != nil {
-			return err
-		}
+	return filepath.Walk(src, walkFn)
+}
 
-		// Chmod it
-		return os.Chmod(dstPath, info.Mode())
+// copyFile copies the contents and mode of the file from src to dst.
+func copyFile(dst, src string, mode os.FileMode) error {
+	srcF, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open source file %q: %w", src, err)
 	}
+	defer srcF.Close()
 
-	return filepath.Walk(src, walkFn)
+	dstF, err := os.Create(dst)
+	if err != nil {
+		return fmt.Errorf("failed to create destination file %q: %w", dst, err)
+	}
+	defer dstF.Close()
```

Comment:
@abstractionfactory fair enough.  I have updated the `copyFile` function to handle potential errors returned by `Close()` on `dstF` to ensure proper error handling on filesystems with deferred errors. If you could please review it would be great.